In [1]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 10: Elves Look, Elves Say ---</h2><p>Today, the Elves are playing a game called <a href="https://en.wikipedia.org/wiki/Look-and-say_sequence">look-and-say</a>.  They take turns making sequences by reading aloud the previous sequence and using that reading as the next sequence.  For example, <code>211</code> is read as "one two, two ones", which becomes <code>1221</code> (<code>1</code> <code>2</code>, <code>2</code> <code>1</code>s).</p>
<p>Look-and-say sequences are generated iteratively, using the previous value as input for the next step.  For each step, take the previous value, and replace each run of digits (like <code>111</code>) with the number of digits (<code>3</code>) followed by the digit itself (<code>1</code>).</p>
<p>For example:</p>
<ul>
<li><code>1</code> becomes <code>11</code> (<code>1</code> copy of digit <code>1</code>).</li>
<li><code>11</code> becomes <code>21</code> (<code>2</code> copies of digit <code>1</code>).</li>
<li><code>21</code> becomes <code>1211</code> (one <code>2</code> followed by one <code>1</code>).</li>
<li><code>1211</code> becomes <code>111221</code> (one <code>1</code>, one <code>2</code>, and two <code>1</code>s).</li>
<li><code>111221</code> becomes <code>312211</code> (three <code>1</code>s, two <code>2</code>s, and one <code>1</code>).</li>
</ul>
<p>Starting with the digits in your puzzle input, apply this process 40 times.  What is <em>the length of the result</em>?</p>
</article>

<p>Your puzzle input is <code class="puzzle-input">3113322113</code>.</p>


In [2]:
tests = [
    {
        "name": "Example 1",
        "s": "1",
        "expected": "11",
    },
    {
        "name": "Example 2",
        "s": "11",
        "expected": "21",
    },
    {
        "name": "Example 3",
        "s": "21",
        "expected": "1211",
    },
    {
        "name": "Example 4",
        "s": "1211",
        "expected": "111221",
    },
    {
        "name": "Example 5",
        "s": "111221",
        "expected": "312211",
    },
]


def look_and_say_one_step(s: str) -> str:
    count, prev = 1, s[0]
    r = []
    for c in s[1:]:
        if c == prev:
            count += 1
        else:
            r.append(str(count))
            r.append(prev)
            count = 1
        prev = c

    r.append(str(count))
    r.append(prev)

    return "".join(r)


run_tests_params(look_and_say_one_step, tests)


Test Example 1 passed, for look_and_say_one_step.
Test Example 1 passed, for look_and_say_one_step.
Test Example 1 passed, for look_and_say_one_step.
Test Example 1 passed, for look_and_say_one_step.
Test Example 5 passed, for look_and_say_one_step.
Success


In [3]:
def look_and_say_one(s: str, n: int) -> str:
    for i in range(n):
        s = look_and_say_one_step(s)
    return s


assert look_and_say_one("1", 5) == "312211"
assert (
    look_and_say_one("1", 14)
    == "311311222113111231131112132112311321322112111312211312111322212311322113212221"
)

In [4]:
len(look_and_say_one("3113322113", 40))

329356

<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>329356</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Neat, right? You might also enjoy hearing <a href="https://www.youtube.com/watch?v=ea7lJkEhytA">John Conway talking about this sequence</a> (that's Conway of <em>Conway's Game of Life</em> fame).</p>
<p>Now, starting again with the digits in your puzzle input, apply this process <em title="Only because any longer started taking alarmingly long on my test hardware, and I wanted to avoid excluding people.">50</em> times.  What is <em>the length of the new result</em>?</p>
</article>

</main>


In [5]:
len(look_and_say_one("3113322113", 50))

4666278

<link href="style.css" rel="stylesheet"></link>

<main>

</p><article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Neat, right? You might also enjoy hearing <a href="https://www.youtube.com/watch?v=ea7lJkEhytA">John Conway talking about this sequence</a> (that's Conway of <em>Conway's Game of Life</em> fame).</p>
<p>Now, starting again with the digits in your puzzle input, apply this process <em title="Only because any longer started taking alarmingly long on my test hardware, and I wanted to avoid excluding people.">50</em> times.  What is <em>the length of the new result</em>?</p>
</article>
<p>Your puzzle answer was <code>4666278</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2015">return to your Advent calendar</a> and try another puzzle.</p>
<p>Your puzzle input was <code class="puzzle-input">3113322113</code>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22Elves+Look%2C+Elves+Say%22+%2D+Day+10+%2D+Advent+of+Code+2015&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F10&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22Elves+Look%2C+Elves+Say%22+%2D+Day+10+%2D+Advent+of+Code+2015+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F10';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>


In [27]:
from itertools import groupby

from more_itertools import flatten


def look_and_say_itr(s: str, n) -> str:
    l = [int(i) for i in s]
    for _ in range(n):
        l = flatten((len(list(b)), a) for a, b in groupby(l))
    return "".join(map(str, l))


look_and_say_itr("1", 5)

'312211'

In [28]:
%%timeit
len(look_and_say_one("3113322113", 50))

2.04 s ± 85.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [30]:
%%timeit
len(look_and_say_itr("3113322113", 50))

7.42 s ± 66.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
